### ISSUE TITLE
Getting a deprecation warning in Python 3.5.1

### ISSUE THREAD

**Issue by jaredLunde** [llllllllll/cnamedtuple#4]
```
DeprecationWarning: builtin type Record has no __module__ attribute
```

**Comment by llllllllll** [comment]
Can you post the code that is generating the warning?

**Comment by jaredLunde** [comment]
It's for an implementation of a psycopg2 cursor factory. The relevant portion is the last method. 

``` python
from cnamedtuple import namedtuple as nt

from psycopg2.extensions import cursor as _cursor


__all__ = ('CNamedTupleCursor',)


class CNamedTupleCursor(_cursor):
    """ Copyright (C) 2003-2010 Federico Di Gregorio  <fog@debian.org> """
    Record = None

    def execute(self, query, vars=None):
        self.Record = None
        return super(CNamedTupleCursor, self).execute(query, vars)

    def executemany(self, query, vars):
        self.Record = None
        return super(CNamedTupleCursor, self).executemany(query, vars)

    def callproc(self, procname, vars=None):
        self.Record = None
        return super(CNamedTupleCursor, self).callproc(procname, vars)

    def fetchone(self):
        t = super(CNamedTupleCursor, self).fetchone()
        if t is not None:
            nt = self.Record
            if nt is None:
                nt = self.Record = self._make_nt()
            return nt._make(t)

    def fetchmany(self, size=None):
        ts = super(CNamedTupleCursor, self).fetchmany(size)
        nt = self.Record
        if nt is None:
            nt = self.Record = self._make_nt()
        return list(map(nt._make, ts))

    def fetchall(self):
        ts = super(CNamedTupleCursor, self).fetchall()
        nt = self.Record
        if nt is None:
            nt = self.Record = self._make_nt()
        return list(map(nt._make, ts))

    def __iter__(self):
        it = super(CNamedTupleCursor, self).__iter__()
        t = next(it)

        nt = self.Record
        if nt is None:
            nt = self.Record = self._make_nt()

        yield nt._make(t)

        while 1:
            yield nt._make(next(it))

    def _make_nt(self):
        return nt("Record", list(map(lambda d: d[0], self.description or [])))
```

**Comment by llllllllll** [comment]
I just built against 3.5 and I am seeing this warning now even with trivial calls. I believe I know what the cause is. This warning should be safe to ignore for now because `__module__` is dynamically added after the type is created.

Also, there is no need to strict the map in the call to `nt` on the last line, you can pass an iterator.

**Comment by jaredLunde** [comment]
Excellent! Thanks for the quick response and the information regarding the map/list.

**Comment by llllllllll** [comment]
np, you might also want to try implicit super. In python 3 `super(MyClass, self)` is the same as `super()` with no arguments. I personally find it much more clear.

**Comment by llllllllll** [comment]
This was sort of tricky. The issue is that it `PyModule_FromSpec` wanted the `PyMethod_Spec` name to be the dotted path to the object. Unfortunatly it also assumes this memory is statically allocated or managed externally. To get around this we need to first create a string object that is the result of `'.'.join((modname, typename))` and then set the `nametuple_spec.name` field to point to the memory for this string object. We then create the type object with this name which will split it on the '.' and create the `__module__` and `__name__` for us. We then need to set `spec.name` back to `""`, free our joined string, and then set the `tp_name` field of our structure to point at the memory allocated for the `ht_name` string of the new type structure. This should make the string lifetimes sane and sharing the memory for the `ht_name` and `tp_name` happens with normal heaptypes so it is somewhat standard. Can you please try out the code on master to make sure that it fixes your issue and doesn't cause any other problems? If everything checks out I will upload the new version to PyPI asap.

As a sidenote, thank you for pinging me on this project. I haven't looked at this in a very long time and it is the first C extension I wrote. Because of that I noticed a lot of reference issues and memory problems, especially around the error handling cases which are harder to test. I did a pretty serious walthrough today and cleaned up everythink I could. This might be a slight performance hit but some might be improvements. The existing benchmarks look about the same, but I am on a newer machine so its all a little faster. A big change is that I realized that the descrwrappers don't need to participate in the cyclical garbage collector because there cannot be a cycle between the descwrapper and the object it is wrapping. It only ever holds a tuple of strings. Named element access is still as fast (and correct) as before so that shouldn't be a concern. Hopefully your project is going well other than this deprecation warning!

**Comment by jaredLunde** [comment]
Very interesting - benching between the two, the difference seems to be pretty negligible on my VM and  the deprecation warning has indeed gone away. Creating the namedtuple is still about ~3200% faster than collections.namedtuple and creating a namedtuple + performing lookup on the namedtuple 1000 times is about 750-800% faster for <10 element namedtuples.

One thing I did notice playing with the cursor, however, is that your implementation doesn't fail like namedtuple does when it gets a field name which shouldn't be allowed (in some cases).

By example:

``` python
from cnamedtuple import namedtuple as cnamedtuple
from collections import namedtuple as stdnamedtuple

# Fails to raise ValueError
Foo_c = cnamedtuple('Foo_c', ('bingo', 'bango', '?bongo'))
Foo_c = cnamedtuple('Foo_c', ('bingo', 'bango', '/bongo'))
Foo_c = cnamedtuple('Foo_c', ('bingo', 'bango', '-bongo'))
# Raises ValueError
Foo_std = stdnamedtuple('Foo_std', ('bingo', 'bango', '?bongo'))
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
#   File "/usr/local/lib/python3.5.1/lib/python3.5/collections/__init__.py", line 384, in namedtuple
#     'identifiers: %r' % name)
# ValueError: Type names and field names must be valid identifiers: '?bongo'
Foo_std = stdnamedtuple('Foo_std', ('bingo', 'bango', '/bongo'))
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
#   File "/usr/local/lib/python3.5.1/lib/python3.5/collections/__init__.py", line 384, in namedtuple
#     'identifiers: %r' % name)
# ValueError: Type names and field names must be valid identifiers: '/bongo'
Foo_std = stdnamedtuple('Foo_std', ('bingo', 'bango', '-bongo'))
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
#   File "/usr/local/lib/python3.5.1/lib/python3.5/collections/__init__.py", line 384, in namedtuple
#     'identifiers: %r' % name)
# ValueError: Type names and field names must be valid identifiers: '-bongo'
```

``` python
# Correctly raises ValueError
Foo_c = cnamedtuple('Foo_c', ('bingo', 'bango', '_bongo'))
Foo_c = cnamedtuple('Foo_c', ('bingo', 'bango', '?bongo?'))
Foo_c = cnamedtuple('Foo_c', ('bingo', 'bango', '/bongo/'))
Foo_c = cnamedtuple('Foo_c', ('bingo', 'bango', '-bongo-'))
```

Happy to help! I really like this project as it has drastically improved the performance of loading large CSVs to namedtuples for me in the past.

**Comment by llllllllll** [comment]
The check was:

``` c
    idx = PyUnicode_GET_LENGTH(field);
    while (--idx) {
        ch = PyUnicode_READ(kind, data, idx);
        if (!(Py_UNICODE_ISALNUM(ch) || ch == (Py_UCS4) '_')) {
            /* Must be all alphanumeric characters or underscores. */
            return CHECKFIELD_NONALNUM;
        }
    }
```

this needs to be `idx--`.

**Comment by llllllllll** [comment]
The error this emits is worded slightly differently. I decided to explicitly say what a valid identifier was here: 

```
ValueError: Type names and field names can only contain alphanumeric characters and underscores: ?bongo
```

do you think this slight difference is okay?

edit: pushed a fix that correctly fails on all of these cases when rename=False

**Comment by jaredLunde** [comment]
I like your exception way better for people who don't have a lot of experience with namedtuple types - it's definitely more explicit than the standard lib. 

I personally don't think it's an issue that the exception text isn't a direct drop-in of the standard lib, although I understand why some would argue that it should be.

Edit:
Also, I've tested the master branch just now and the naming issue is indeed fixed.

**Comment by llllllllll** [comment]
Yeah, I am trying to make this as drop in of a replacement as possible. The only thing I am really okay with is losing the `_source` attribute because there is no python source for the class. Idle thought, it might be funny to make it a reference to the c code for cnamedtuple. Also `verbose` is not practical for anything but the repl, and even then it is mostly for a teaching tool. I will try to upload these fixes to pypi tonight

**Comment by jaredLunde** [comment]
Reading the thread on this (http://bugs.python.org/issue19640), `verbose` seems to be outdated due to the `_source` attribute, and the reason neither have been removed is because the author didn't see the use in saving 2k per namedtuple for the added complexity of making it a lazy evaluation. 

As these two options have zero use to your implementation and there was serious consideration about removing it from the standard library, I personally can't see any good reason why you shouldn't just do away with them.

Thinking along this same line, it may even be detrimental to keep the `_source` attribute if people are trying to use it for any practical purpose. Better to raise an `AttributeError` than to encounter some other type of unknown error in the process of trying to utilize the attribute.

**Comment by llllllllll** [comment]
I have pushed these fixes up to pypi as cnamedtuple 0.1.6. Thank you for letting me know about this!
